--- a/net/minecraft/world/level/ServerExplosion.java
+++ b/net/minecraft/world/level/ServerExplosion.java
@@ -88,7 +_,7 @@
             }
         }
 
-        CACHED_RAYS = rayCoords.toDoubleArray();
+        CACHED_RAYS = sortExplosionRays(rayCoords); // Sakura - optimise paper explosions
     }
 
     private static final int CHUNK_CACHE_SHIFT = 2;
@@ -296,7 +_,12 @@
                             Math.fma(dz, diffZ, offZ)
                     );
 
-                    if (!this.clipsAnything(from, source, context, blockCache, blockPos)) {
+                    // Sakura start - replace explosion density cache
+                    final float density = this.level.densityCache.getKnownDensity(from);
+                    if (density != me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
+                        missedRays += (int) density;
+                    } else if (!this.clipsAnything(from, source, context, blockCache, blockPos)) {
+                        // Sakura end - replace explosion density cache
                         ++missedRays;
                     }
                 }
@@ -306,6 +_,55 @@
         return (float)missedRays / (float)totalRays;
     }
     // Paper end - collisions optimisations
+    // Sakura start - optimise paper explosions
+    /*
+     * Sort the explosion rays to better utilise the chunk and block cache.
+     *   x +   Vanilla     Sorted
+     * z @ z      8           5
+     * - x      6   7       6   4
+     *        4   @   5   7   @   3
+     *          2   3       8   2
+     *            1           1
+     */
+    private static double[] sortExplosionRays(it.unimi.dsi.fastutil.doubles.DoubleArrayList rayCoords) {
+        List<double[]> explosionRays = new ArrayList<>();
+
+        for (int i = 0; i < rayCoords.size(); i += 3) {
+            double[] ray = new double[3];
+            rayCoords.getElements(i, ray, 0, 3);
+            explosionRays.add(ray);
+        }
+
+        rayCoords.clear();
+        explosionRays.sort(java.util.Comparator.comparingDouble(vec -> {
+            double sign = Math.signum(vec[0]);
+            double dir = (sign - 1) / 2;
+            return sign + 8 + vec[2] * dir;
+        }));
+
+        double[] rays = new double[explosionRays.size() * 3];
+        for (int i = 0; i < explosionRays.size() * 3; i++) {
+            rays[i] = explosionRays.get(i / 3)[i % 3];
+        }
+        return rays;
+    }
+    // Sakura end - optimise paper explosions
+    // Sakura start - optimise explosion protected regions
+    protected final boolean isRegionUnprotected() {
+        // optimisation: We check if a plugin has cancelled the event or cleared the blockList.
+        // It tells us if the result was thrown away, so we can avoid the block searching logic.
+        // As a side effect the event is called twice which may interfere with some plugins.
+        if (this.source != null && this.level.sakuraConfig().cannons.explosion.optimiseProtectedRegions) {
+            Location location = new Location(this.level.getWorld(), this.center.x, this.center.y, this.center.z);
+            List<org.bukkit.block.Block> blocks = new ObjectArrayList<>(1);
+            blocks.add(location.getBlock());
+            EntityExplodeEvent event = CraftEventFactory.callEntityExplodeEvent(this.source, blocks, 0.0f, this.blockInteraction);
+            return !event.isCancelled() && !event.blockList().isEmpty();
+        }
+
+        return true;
+    }
+    // Sakura end - optimise explosion protected regions
 
     public ServerExplosion(
         ServerLevel level,
@@ -350,8 +_,17 @@
                         double d9 = Mth.lerp(d6, boundingBox.minY, boundingBox.maxY);
                         double d10 = Mth.lerp(d7, boundingBox.minZ, boundingBox.maxZ);
                         Vec3 vec3 = new Vec3(d8 + d3, d9, d10 + d4);
-                        if (entity.level().clip(new ClipContext(vec3, explosionVector, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType()
-                            == HitResult.Type.MISS) {
+
+                        // Sakura start - replace explosion density cache
+                        final net.minecraft.world.phys.HitResult.Type hitResult;
+                        final float density = entity.level().densityCache.getKnownDensity(vec3);
+                        if (density != me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
+                            hitResult = density != 0.0f ? net.minecraft.world.phys.HitResult.Type.MISS : net.minecraft.world.phys.HitResult.Type.BLOCK;
+                        } else {
+                            hitResult = entity.level().clip(new ClipContext(vec3, explosionVector, ClipContext.Block.COLLIDER, ClipContext.Fluid.NONE, entity)).getType();
+                        }
+                        if (hitResult == HitResult.Type.MISS) {
+                            // Sakura end - replace explosion density cache
                             i++;
                         }
 
@@ -396,6 +_,17 @@
             initialCache = this.getOrCacheExplosionBlock(blockX, blockY, blockZ, key, true);
         }
 
+        // Sakura start - optimise paper explosions
+        if (!this.interactsWithBlocks() || initialCache.resistance > (this.radius * 1.3f)) {
+            return ret;
+        }
+        // Sakura end - optimise paper explosions
+        // Sakura start - optimise protected explosions
+        if (!this.isRegionUnprotected()) {
+            return ret;
+        }
+        // Sakura end - optimise protected explosions
+
         // only ~1/3rd of the loop iterations in vanilla will result in a ray, as it is iterating the perimeter of
         // a 16x16x16 cube
         // we can cache the rays and their normals as well, so that we eliminate the excess iterations / checks and
@@ -475,16 +_,55 @@
         // Paper end - collision optimisations
     }
 
-    private void hurtEntities() {
-        float f = this.radius * 2.0F;
+    // Sakura start - optimise paper explosions
+    protected final AABB getExplosionBounds(float f) {
         int floor = Mth.floor(this.center.x - f - 1.0);
         int floor1 = Mth.floor(this.center.x + f + 1.0);
         int floor2 = Mth.floor(this.center.y - f - 1.0);
         int floor3 = Mth.floor(this.center.y + f + 1.0);
         int floor4 = Mth.floor(this.center.z - f - 1.0);
         int floor5 = Mth.floor(this.center.z + f + 1.0);
-        List <Entity> list = this.level.getEntities(excludeSourceFromDamage ? this.source : null, new AABB(floor, floor2, floor4, floor1, floor3, floor5), entity -> entity.isAlive() && !entity.isSpectator()); // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source
-        for (Entity entity : list) { // Paper - used in loop
+        return new AABB(floor, floor2, floor4, floor1, floor3, floor5);
+    }
+
+    private void hurtEntities() {
+        float f = this.radius * 2.0F;
+
+        int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection(this.level);
+        int maxSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMaxSection(this.level);
+
+        int minChunkX = Mth.floor(this.center.x - f) >> 4;
+        int maxChunkX = Mth.floor(this.center.x + f) >> 4;
+        int minChunkY = Mth.clamp(Mth.floor(this.center.y - f) >> 4, minSection, maxSection);
+        int maxChunkY = Mth.clamp(Mth.floor(this.center.y + f) >> 4, minSection, maxSection);
+        int minChunkZ = Mth.floor(this.center.z - f) >> 4;
+        int maxChunkZ = Mth.floor(this.center.z + f) >> 4;
+
+        ca.spottedleaf.moonrise.patches.chunk_system.level.entity.EntityLookup entityLookup = this.level.moonrise$getEntityLookup();
+        for (int chunkX = minChunkX; chunkX <= maxChunkX; ++chunkX) {
+            for (int chunkZ = minChunkZ; chunkZ <= maxChunkZ; ++chunkZ) {
+                ca.spottedleaf.moonrise.patches.chunk_system.level.entity.ChunkEntitySlices chunk = entityLookup.getChunk(chunkX, chunkZ);
+                if (chunk == null) continue; // empty slice
+
+                for (int chunkY = minChunkY; chunkY <= maxChunkY; ++chunkY) {
+                    this.impactEntities(f, chunk.getSectionEntities(chunkY));
+                }
+            }
+        }
+    }
+
+    protected final void impactEntities(float f, Entity[] entities) {
+        for (int i = 0; i < entities.length; i++) {
+            Entity entity = entities[i];
+            if (entity == null) break;      // end of entity section
+            this.impactEntity(f, entity);
+            if (entity != entities[i]) i--; // entities can be removed mid-explosion
+        }
+    }
+
+    protected final void impactEntity(float f, Entity entity) {
+        if (entity.isAlive() && !entity.isSpectator() && (!this.excludeSourceFromDamage || entity != this.source)) { // Paper - Fix lag from explosions processing dead entities, Allow explosions to damage source
+            // Sakura end - optimise paper explosions
             if (!entity.ignoreExplosion(this)) {
                 double d = Math.sqrt(entity.distanceToSqr(this.center)) / f;
                 if (d <= 1.0) {
@@ -509,15 +_,16 @@
                             // - Damaging EntityEnderDragon does nothing
                             // - EnderDragon hitbock always covers the other parts and is therefore always present
                             if (entity instanceof EnderDragonPart) {
-                                continue;
+                                return; // Sakura - optimise paper explosions
                             }
 
                             entity.lastDamageCancelled = false;
 
                             if (entity instanceof EnderDragon) {
+                                AABB bounds = this.getExplosionBounds(f); // Sakura - optimise paper explosions
                                 for (EnderDragonPart dragonPart : ((EnderDragon) entity).getSubEntities()) {
                                     // Calculate damage separately for each EntityComplexPart
-                                    if (list.contains(dragonPart)) {
+                                    if (dragonPart.getBoundingBox().intersects(bounds)) { // Sakura - optimise paper explosions
                                         dragonPart.hurtServer(this.level, this.damageSource, this.damageCalculator.getEntityDamageAmount(this, entity, f1));
                                     }
                                 }
@@ -526,7 +_,7 @@
                             }
 
                             if (entity.lastDamageCancelled) { // SPIGOT-5339, SPIGOT-6252, SPIGOT-6777: Skip entity if damage event was cancelled
-                                continue;
+                                return; // Sakura - optimise paper explosions
                             }
                             // CraftBukkit end
                         }
@@ -610,6 +_,11 @@
             return;
         }
         // CraftBukkit end
+        // Sakura start - replace explosion density cache
+        if (!blocks.isEmpty() && !this.level.paperConfig().environment.optimizeExplosions) {
+            this.level.densityCache.invalidate();
+        }
+        // Sakura end - replace explosion density cache
 
         for (BlockPos blockPos : blocks) {
             // CraftBukkit start - TNTPrimeEvent
@@ -770,14 +_,12 @@
 
     // Paper start - Optimize explosions
     protected float getBlockDensity(Vec3 vec3d, Entity entity) {
-        if (!this.level.paperConfig().environment.optimizeExplosions) {
-            return this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
-        }
-        CacheKey key = new CacheKey(this, entity.getBoundingBox());
-        Float blockDensity = this.level.explosionDensityCache.get(key);
-        if (blockDensity == null) {
+        // Sakura start - replace explosion density cache
+        float blockDensity = this.level.densityCache.getDensity(vec3d, entity);
+        if (blockDensity == me.samsuik.sakura.explosion.density.BlockDensityCache.UNKNOWN_DENSITY) {
             blockDensity = this.getSeenFraction(vec3d, entity, this.directMappedBlockCache, this.mutablePos); // Paper - collision optimisations
-            this.level.explosionDensityCache.put(key, blockDensity);
+            this.level.densityCache.putDensity(vec3d, entity, blockDensity);
+            // Sakura end - replace explosion density cache
         }
 
         return blockDensity;
