From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Hermelijn15 <j.s.dekker-1@student.utwente.nl>
Date: Fri, 17 Nov 2023 20:09:03 +0100
Subject: [PATCH] Added list of ItemEntity's that ignore explosions


diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index 0f086af57a5ff08c264dcbf89a8c3931ec73a609..d9a6324eeb0cfaef3d0bbff884ddfadc6e4132f7 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -410,6 +410,11 @@ public class ItemEntity extends Entity implements TraceableEntity {
 
     @Override
     public boolean ignoreExplosion(Explosion explosion) {
+        // Sakura start - add list of items that ignore explosions
+        if (this.level().sakuraConfig().entity.items.explosionResistantItems.contains(this.getItem().getItem())) {
+            return true;
+        }
+        // Sakura end - add list of items that ignore explosions
         return explosion.shouldAffectBlocklikeEntities() ? super.ignoreExplosion(explosion) : true;
     }
 
