From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Samsuik <40902469+Samsuik@users.noreply.github.com>
Date: Fri, 8 Dec 2023 18:21:56 +0000
Subject: [PATCH] Add instant mob death animation


diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index e7603a67c985ff105a8a4e16179fe33fe26f5b22..1bd9f31da1ec66259dc6391448e2b8ce69ddb817 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -1869,6 +1869,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
 
             // Paper start
             if (this.dead) { // Paper
+                // Sakura start
+                if (level().sakuraConfig().entity.instantDeathAnimation && !(this instanceof Player)) {
+                    this.deathTime = 20;
+                    return;
+                }
+                // Sakura end
                 this.level().broadcastEntityEvent(this, (byte) 3);
 
             this.setPose(Pose.DYING);
