--- a/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
+++ b/ca/spottedleaf/moonrise/patches/chunk_system/level/entity/EntityLookup.java
@@ -722,6 +_,13 @@
 
     public void getEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate,
                             final int maxCount) {
+        // Sakura start - add maxSearch to getEntities
+        this.getEntities(except, box, into, predicate, maxCount, Integer.MAX_VALUE);
+    }
+
+    public void getEntities(final Entity except, final AABB box, final List<Entity> into, final Predicate<? super Entity> predicate,
+                            final int maxCount, final int maxSearch) {
+        // Sakura end - add maxSearch to getEntities
         final int minChunkX = (Mth.floor(box.minX) - 2) >> 4;
         final int minChunkZ = (Mth.floor(box.minZ) - 2) >> 4;
         final int maxChunkX = (Mth.floor(box.maxX) + 2) >> 4;
@@ -753,7 +_,7 @@
                             continue;
                         }
 
-                        if (chunk.getEntities(except, box, into, predicate, maxCount)) {
+                        if (chunk.getEntities(except, box, into, predicate, maxCount, maxSearch)) { // Sakura - add maxSearch to getEntities
                             return;
                         }
                     }
