--- a/src/main/java/org/bukkit/command/defaults/VersionCommand.java
+++ b/src/main/java/org/bukkit/command/defaults/VersionCommand.java
@@ -33,6 +_,11 @@
 import net.kyori.adventure.text.format.TextDecoration;
 import net.kyori.adventure.text.serializer.plain.PlainTextComponentSerializer;
 // Paper end - version command 2.0
+// Sakura start
+import net.kyori.adventure.text.event.HoverEvent;
+import net.kyori.adventure.text.minimessage.MiniMessage;
+import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;
+// Sakura end
 
 public class VersionCommand extends BukkitCommand {
     private VersionFetcher versionFetcher; // Paper - version command 2.0
@@ -44,6 +_,15 @@
         return versionFetcher;
     }
 
+    // Sakura start
+    private static final String VERSION_MESSAGE = """
+                <dark_purple>.
+                <dark_purple>| <white>This server is running <gradient:red:light_purple>Sakura</gradient>
+                <dark_purple>| <white>Commit<dark_gray>: \\<<commit>> <gray>targeting </gray>(<yellow>MC</yellow>: <gray><version></gray>)
+                <dark_purple>| <white>Github<dark_gray>: \\<<yellow><click:open_url:'https://github.com/Samsuik/Sakura'>link</click></yellow>>
+                <dark_purple>'""";
+    // Sakura end
+
     public VersionCommand(@NotNull String name) {
         super(name);
 
@@ -55,12 +_,18 @@
 
     @Override
     public boolean execute(@NotNull CommandSender sender, @NotNull String currentAlias, @NotNull String[] args) {
-        if (!testPermission(sender)) return true;
+        // Sakura - move down into else
 
         if (args.length == 0) {
             //sender.sendMessage("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")"); // Paper - moved to setVersionMessage
-            sendVersion(sender);
-        } else {
+            // Sakura start
+            sender.sendMessage(MiniMessage.miniMessage().deserialize(VERSION_MESSAGE,
+                Placeholder.component("commit", Component.text("hover", NamedTextColor.YELLOW)
+                    .hoverEvent(HoverEvent.showText(Component.text(Bukkit.getGitInformation())))),
+                Placeholder.unparsed("version", Bukkit.getMinecraftVersion())
+            ));
+        } else if (testPermission(sender)) {
+            // Sakura end
             StringBuilder name = new StringBuilder();
 
             for (String arg : args) {
